Fix bug 1725648: Add project insights

eemeli opened the pull request:
This does a bunch of things in order to enable the "Insights" tab to work for projects:
- The `ProjectInsightsSnapshot` class/model is expanded to have the same data fields as `LocaleInsightsSnapshot` does. The "Active users" and "Unreviewed lifespan" values are not yet used for anything, though.
- In addition to Django's auto-generated corresponding db migrations, a couple of manually crafted SQL scripts are included to fill the historical data, starting from the current 2020-12-24 origin of project insight data. Running these should take at most a few seconds.
- The daily-run `collect_project_insights()` task is expanded to fill the db with the expected data. A little bit of reorg is done to make sure that code or db calls aren't duplicated with the `collect_locale_insights()` task.
- If `settings.ENABLE_INSIGHTS_TAB` is on, the UI includes the Insights tab and provides it from `/projects/:slug/insights`. The necessary dependencies are added to `project.min.js` and `project.min.css`.
- A new ajax endpoint `/projects/:slug/ajax/insights` is added, providing the data.
- The Django templates and JS scripts providing the actual insight contents are slightly modified, such that they are resilient to getting partial data only, as project insights do not (yet) include the "Active users" and "Unreviewed suggestions lifespan" data, and for the rest have it available for 8 months at most (rather than a full year).

Once this is accepted, I'll be able to file a subsequent PR for [bug 1725649](https://bugzilla.mozilla.org/show_bug.cgi?id=1725649), adding insights for the project+locale combination with a similar shape. If we subsequently expand these insights to also include the full data that's provided for locale insights, then after that, sometime early next year, it'll make sense to normalise these three views to all use the project+locale data rather than ~triplicating the work. But not yet, because we'd needlessly lose a few months of data that we currently provide for locale insights.

One possibility is to already now drop the `ProjectInsightsSnapshot`, and only update `ProjectLocaleInsightsSnapshot` & use its data for both project & project+locale insights. Or I could do that in the above-mentioned follow-on PR, if we don't mind the intermediate db migration that's included here.

eemeli commented on pontoon/insights/migrations/0003_project_insights.py:
```diff
@@ -0,0 +1,199 @@
+import pontoon.insights.models
+
+
+_activity_sql = """
```
Note that those instructions are specific to `RunPython`, and this migration is using `RunSQL`. If you still think it's beneficial, I can split the migration into two steps.

mathjazz replied:
Sounds like we're good.

mathjazz replied:
I just tried to unapply the migration on stage and hit:
```
django.db.utils.OperationalError: cannot ALTER TABLE "insights_projectinsightssnapshot" because it has pending trigger events
```

Interestingly enough, this is the error Django docs talk about when combining schema changes and `RunPython` operations. Perhaps `RunSQL` doesn't make a difference?

*On databases that do support DDL transactions (SQLite and PostgreSQL), RunPython operations do not have any transactions automatically added besides the transactions created for each migration. Thus, on PostgreSQL, for example, you should avoid combining schema changes and RunPython operations in the same migration or you may hit errors like OperationalError: cannot ALTER TABLE "mytable" because it has pending trigger events.*

eemeli replied:
Okay, I'll separate the data migrations into their own thing.

eemeli commented on pontoon/insights/migrations/0003_project_insights.py:
```diff
@@ -0,0 +1,199 @@
+  human_translations = sums.human_translations,
+  machinery_translations = sums.machinery_translations,
+  new_suggestions = sums.new_suggestions,
+  self_approved = sums.self_approved,
```
@mathjazz Given that this is a one-off operation, does it really make sense to rewrite it using Django ORM? My suspicion while working on this was that the views of the data that are being used by this `UPDATE` might require some custom ORM models to be used to represent the data, and that it would indeed be quite a bit more work to represent the data there. The performance difference is probably measurable, but irrelevant as this only needs to be done once.

@jotes I have not experimented yet with window functions. But they do look rather interesting!

mathjazz replied:
`F()` and `Func()` sound even less readable. :) So let's stick to SQL, especially now that we already have it.

eemeli commented on pontoon/insights/static/js/insights.js:
```diff
@@ -210,38 +216,38 @@ var Pontoon = (function (my) {
                                 hidden: true,
                             },
-                        ],
+                        ].filter(Boolean),
```
It's common in my experience? As in, I've seen it in code written by others on multiple occasions. And I gather from your comment that it was at least eventually decipherable?

mathjazz replied:
It was decipherable as anything else is after a little bit of Googling. :)

The reason I'd prefer a comment is to save others from needing to do the same.

eemeli commented on pontoon/insights/migrations/0003_project_insights.py:
```diff
@@ -0,0 +1,199 @@
+      (action_type = 'translation:created' AND act.performed_by_id = tra.approved_user_id)
+    ) AS self_approved,
+    action_type = 'translation:approved' AND act.performed_by_id != tra.user_id AS peer_approved,
+    action_type = 'translation:rejected' AS rejected
```
I think that's a question that ought to be addressed separately from this PR, as this is just replicating the current logic: https://github.com/mozilla/pontoon/blob/eebb312e4a217adec66dafa94848fb1979260d3f/pontoon/insights/tasks.py#L454-L455

eemeli commented on pontoon/insights/utils.py:
```diff
@@ -1,20 +1,28 @@
+    return datetime.now() - relativedelta(months=12)
+
+
+def get_locale_insights(query_filters=None):
```
How about if I normalise in the next step the interface to `get_insights(locale, project)`, and do the right thing if only one or both parameters are given?

mathjazz replied:
SGTM

mathjazz commented:
@flodolo IIRC you have some historic data for some projects plotted on charts. Could you please share the links so that we can see how that compares to charts generated by this patch on stage?

flodolo commented:
> @flodolo IIRC you have some historic data for some projects plotted on charts.

I only have it for Firefox (you might want to filter to one locale, or a group of locales, to make sense of the graph)
https://test.flod.org/firefox-l10n/

eemeli commented:
@mathjazz I think all of the previously raised concerns have now been addressed?

mathjazz commented:
Thanks for the update. Taking it for one last spin on stage with latest data...

mathjazz commented:
It seems like the **New source string** count isn't always what I would expect it to be.

I used the following script to collect the expected numbers for Firefox and compared them with the numbers on the [chart](https://mozilla-pontoon-staging.herokuapp.com/projects/firefox/insights/).

```python
from pontoon.base.utils import aware_datetime
from pontoon.base.models import *

Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2020,12,1), date_created__lt=aware_datetime(2021,1,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,1,1), date_created__lt=aware_datetime(2021,2,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,2,1), date_created__lt=aware_datetime(2021,3,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,3,1), date_created__lt=aware_datetime(2021,4,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,4,1), date_created__lt=aware_datetime(2021,5,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,5,1), date_created__lt=aware_datetime(2021,6,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,6,1), date_created__lt=aware_datetime(2021,7,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,7,1), date_created__lt=aware_datetime(2021,8,1)).count()
Entity.objects.filter(resource__project__slug="firefox", date_created__gte=aware_datetime(2021,8,1), date_created__lt=aware_datetime(2021,9,1)).count()
```

Numbers for the following months are different:
- **December**: 0 new source strings are shown on the chart instead of the expected 95.
- **March** and **April**: probably a timezone issue, since 49 entities submitted on March 31 UTC are moved to April.

mathjazz commented:
One more issue - 13 months are now shown on the team Insights dashboard:

https://pontoon.mozilla.org/fr/insights/
https://mozilla-pontoon-staging.herokuapp.com/fr/insights/

mathjazz commented:
And, since the absolute numbers (in tooltips) are now higher (than 1000), it would be great to add some number formatting (in particular a thousands separator).

eemeli commented:
The December 2020 numbers are different because the insights data is generated only from 2020-12-24 onwards, and all of December's new sources came earlier in the month.

The March/April disparity is actually intentional, as it's made to match the existing behaviour of the locale-specific insights data: The rows are filled with data from the day immediately before the `created_at` date. For example, if you look at the previously generated locale-specific `new_source_strings` data, you'll see the matching entries all logged for 2021-04-01 rather than 2021-03-31.

Regarding the insights, I don't see 13 months in the staging graphs but I do see an off-by-one difference. Will investigate.

mathjazz commented:
> The December 2020 numbers are different because the insights data is generated only from 2020-12-24 onwards, and all of December's new sources came earlier in the month.

Phew. Good point. I think this will be confusing, though. Even if just for a few months. What if we start with 2020-12-01? Does it matter that we only started collecting stats on 2020-12-24? They are only shown as averages if I'm not mistaken, so we could just use the values of 2020-12-24 for the first 23 days of December. Alternatively, we should use 2021-01-01 as the start date.

> The March/April disparity is actually intentional, as it's made to match the existing behaviour of the locale-specific insights data: The rows are filled with data from the day immediately before the `created_at` date. For example, if you look at the previously generated locale-specific `new_source_strings` data, you'll see the matching entries all logged for 2021-04-01 rather than 2021-03-31.

Ah, sorry.

> Regarding the insights, I don't see 13 months in the staging graphs but I do see an off-by-one difference. Will investigate.

Yeah, I'm sure that was the case yesterday (Aug 31 vs Sep 1).

jotes commented:
I had only explained the reasoning behind the suggestion (i  was assuming
that we'll end up with ru python calls). If we stick to runSQL then there's
no need for split this.

On Sunday, August 29, 2021, Eemeli Aro ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pontoon/insights/migrations/0003_project_insights.py
> <https://github.com/mozilla/pontoon/pull/1990#discussion_r698007884>:
>
> > @@ -0,0 +1,199 @@
> +# Generated by Django 3.2.4 on 2021-08-24 06:45
> +
> +import datetime
> +from django.db import migrations, models
> +import pontoon.insights.models
> +
> +
> +_activity_sql = """
>
> Note that those instructions are specific to RunPython, and this
> migration is using RunSQL. If you still think it's beneficial, I can
> split the migration into two steps.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla/pontoon/pull/1990#discussion_r698007884>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDAYTG5NN74K5GFRFSUXVDT7IRRHANCNFSM5CY3XWMA>
> .
>


-- 
Kind regards

Jarek "jotes" Śmiejczak, a python programmer by day, a mozillian by heart...
Homepage: http://jotes.work, Github: http://github.com/jotes
Mobile: +48693027040, Mozillian: http://mozillians.org/u/jotes

eemeli commented:
> > The December 2020 numbers are different because the insights data is generated only from 2020-12-24 onwards, and all of December's new sources came earlier in the month.
> 
> Phew. Good point. I think this will be confusing, though. Even if just for a few months. What if we start with 2020-12-01? Does it matter that we only started collecting stats on 2020-12-24? They are only shown as averages if I'm not mistaken, so we could just use the values of 2020-12-24 for the first 23 days of December. Alternatively, we should use 2021-01-01 as the start date.

I think the easiest fix that would not lose current data or require generating fake data would be to keep the insights data as currently from 2020-12-24 onwards, but add a filter to the query to only select from 2021 when generating the ajax response. Would that be okay?

mathjazz commented:
> I think the easiest fix that would not lose current data or require generating fake data would be to keep the insights data as currently from 2020-12-24 onwards, but add a filter to the query to only select from 2021 when generating the ajax response. Would that be okay?

Let's do that.

eemeli commented:
@mathjazz I fixed the start date logic and switched the code to read the months from the actual data, rather than recalculating them. This should clear up edge cases around month changes as well as ensuring that the oldest month is getting its full data.

mathjazz commented on pontoon/insights/utils.py:
```diff
@@ -1,27 +1,35 @@
+    so limit queries to start from 2021-01-01 at earliest.
+    """
+    now = datetime.now()
+    if from2021 and now.year == 2021:
```
Mind filing a bug to remove that flag after the new year party?

mathjazz commented on pontoon/insights/static/js/insights.js:
```diff
@@ -274,48 +274,33 @@ var Pontoon = (function (my) {
+                                            return base + '%';
+                                        case 'Human translations':
+                                        case 'Machinery translations': {
+                                            const pct = Pontoon.insights.getPercent(
```
Thanks a bunch, this looks much more readable!

mathjazz reviewed (approved):
Excellent work!

![](https://media.giphy.com/media/JIJTSa0Wsg9fW/giphy.gif?cid=ecf05e47tqk7skc8z2e0xoq8str5pxfm4vntq0r1hx2c9p4o&rid=giphy.gif&ct=g)

SouravVermaCertirfied commented:
that was great.
